src/python_testing/TC_IDM_4_2.py

```diff
@@ -0,0 +1,589 @@
+        sub_cr1_step2 = await CR1.ReadAttribute(
+            nodeid=self.dut_node_id,
+            attributes=node_label_attr_path,
+            reportInterval=(min_interval_floor_sec, max_interval_ceiling_sec),
+            keepSubscriptions=False
+        )
+
+        # Verify attribute data came back
+        self.verify_attribute_data(
+            sub=sub_cr1_step2,
+            cluster=Clusters.BasicInformation,
+            attribute=node_label_attr
+        )
+
+        # Verify subscriptionId is of uint32 type
+        asserts.assert_true(self.is_uint32(sub_cr1_step2.subscriptionId), "subscriptionId is not of uint32 type.")
+
+        # Verify MaxInterval is of uint32 type
+        sub_cr1_step2_intervals = sub_cr1_step2.GetReportingIntervalsSeconds()
+        sub_cr1_step2_min_interval_floor_sec, sub_cr1_step2_max_interval_ceiling_sec = sub_cr1_step2_intervals
+        asserts.assert_true(self.is_uint32(sub_cr1_step2_max_interval_ceiling_sec), "MaxInterval is not of uint32 type.")
+
+        # Verify MaxInterval is less than or equal to MaxIntervalCeiling
+        asserts.assert_true(sub_cr1_step2_max_interval_ceiling_sec <= max_interval_ceiling_sec,
+                            "MaxInterval is not less than or equal to MaxIntervalCeiling")
+
+        sub_cr1_step2.Shutdown()
+
+        '''
+        ##########
+        Step 3
+        ##########
+        '''
+        self.print_step(3, "Setup CR2 such that it does not have access to a specific cluster. CR2 sends a subscription message to subscribe to an attribute on that cluster for which it does not have access.")
+
+        # Limited ACE for controller 2 with single cluster access
+        CR2_limited_ace = Clusters.AccessControl.Structs.AccessControlEntryStruct(
+            privilege=Clusters.AccessControl.Enums.AccessControlEntryPrivilegeEnum.kView,
+            authMode=Clusters.AccessControl.Enums.AccessControlEntryAuthModeEnum.kCase,
+            targets=[Clusters.AccessControl.Structs.AccessControlTargetStruct(cluster=Clusters.BasicInformation.id)],
+            subjects=[CR2_nodeid])
+
+        # Add controller 2 limited ACE to DUT ACL
+        dut_acl = copy.deepcopy(dut_acl_original)
+        dut_acl.append(CR2_limited_ace)
+
+        # Write updated DUT ACL into DUT
+        await CR1.WriteAttribute(
+            nodeid=self.dut_node_id,
+            attributes=[(0, Clusters.AccessControl.Attributes.Acl(dut_acl))]
+        )
+
+        # Controller 2 tries to subscribe an attribute from a cluster
+        # it doesn't have access to
+        # "INVALID_ACTION" status response expected
+        try:
+            await CR2.ReadAttribute(
+                nodeid=self.dut_node_id,
+                # Attribute from a cluster controller 2 has no access to
+                attributes=[(0, Clusters.AccessControl.Attributes.Acl)],
+                keepSubscriptions=False,
+                reportInterval=[3, 3],
+                autoResubscribe=False
+            )
```

Comment:
not throwing seems like an issue here (and below) so you have to add an assert that complains right after this, so if no error is thrown we get a test failure.

Reply:
Updated